pyteal/compiler/subroutines.py

```diff
@@ -107,13 +106,22 @@ def spillLocalSlotsDuringRecursion(
             before: List[TealComponent] = []
             after: List[TealComponent] = []
 
-            if len(reentryPoints.intersection(stmt.getSubroutines())) != 0:
+            reentrySubroutineCalls = reentryPoints.intersection(stmt.getSubroutines())
+            if len(reentrySubroutineCalls) != 0:
                 # A subroutine is being called which may reenter the current subroutine, so insert
                 # ops to spill local slots to the stack before calling the subroutine and also to
                 # restore the local slots after returning from the subroutine. This prevents a
                 # reentry into the current subroutine from modifying variables we are currently
                 # using.
 
+                # the only opcode that references subroutines is callsub, and it should only ever
+                # reference one subroutine at a time
+                assert (
+                    len(reentrySubroutineCalls) == 1
+                ), "Multiple subroutines are called from the same op"
+                reentrySubroutineCall = reentrySubroutineCalls.pop()
```

Comment:
@jasonpaulos For my edification:  Have you considered the following changes to catch the highlighted problem earlier?
* In `TealOp` constructor, validate maximum of 1 `SubroutineDefinition` provided or raise error.
* Change `TealOp.getSubroutines` to something like `Teal.subroutine() -> Optional[SubroutineDefinition]`.

I'm trying to gauge:
* Does the proposed change presents a technical barrier?  From my cursory review, it seems like _no_.
* If no technical barrier, I'm wondering if you had considered the change, but elected for a more smaller scoped fix?  I'm on-board with resolving the bug ASAP though I'm also interested in potential follow on PRs to any agreed upon improvements.

----

Additionally, presuming we keep as shown:  Can you consider the following concern and let me know if you agree with the proposed change?
* PyCharm shows `reentrySubroutineCalls` is a `set`.  https://docs.python.org/3.6/library/stdtypes.html#frozenset.pop shows`set.pop()` removes an _arbitrary_ element.
* While asserting `len(reentrySubroutineCalls) == 1`, `pop()` usage is _safe_. 
* My concern is:
  * I think the approach can be made more robust with moderate effort.  I think it's unlikely we'll recall the Set semantics if/when the assertion is relaxed.
  * There are probably multiple ways to improve robustness and I'm flexible here.  One approach is to convert `reentrySubroutineCalls` to a `list`.

Reply:
@jasonpaulos Thanks for the change in https://github.com/algorand/pyteal/pull/234/commits/07e48b6f5fe4d296b2979ec157028ea00907302f.  

The commit solves my immediate concern and I'll approve the PR.  Though I hold the thread open in case you'd like to add commentary about asserting earlier.

Reply:
@jasonpaulos Gotcha - Thanks for the added context.  I have additional questions that I'll can pursue out-of-bands to the PR.  Resolving.